feat(nuemark): inline span support, inline and block html nuemark tag  support

nobkd opened the pull request:
- incremental pr based on: #476
- fix: #486

Adds basic support for inline spans:
- `hello [.green "*content*"] etc` -> `hello <span class="green"><em>content</em></span> etc`
- `hello [.myclass]` -> `hello <span class="myclass"></span>`

Adds support for native html tags with block and inline nuemark tags:
- `hello [del "content"]` -> `hello <del>content</del>` (inline)
- `[del "content"]` -> `<del>content</del>` (empty block with `_` attribute)
- `[section.sub]\n  hello` -> `<section class="sub"><p>hello</p></section>` (block)

---

to be done:
- [x] inline `span` support, through anonymous tag (`hi [.classname]` -> `hi <span class="classname"></span>`; `hi [.c "hi"]` -> `hi <span class="c">hi</span>`)
- [x] other inline html tags (`[i "content"]` -> `<i>content</i>`) (without manual component)
- [x] other html block tags for block scope nuemark tags (without manual component)
- [x] test custom attributes on nuemark html tags (`a [.hello world "content"]` -> `a <span class="hello" world>content</span>`) 
- [x] fix lone/starting unordered lists for named tags (maybe other pr? See #476)
- [ ] fix native elements having all page data as attributes
- [ ] docs entry about html nuemark tags

nobkd committed:
fix toc cropped, fix heading not hidden in zen mode

nobkd committed:
inline span test

goblinfactory commented:
Are the quotes entirely necessary?

```markdown
hello [.green "*content*"] etc -> hello <span class="green"><em>content</em></span> etc
```

could also be made to work as follows, without requiring quotes

```markdown
hello [.green *content*] etc -> hello <span class="green"><em>content</em></span> etc
```

Which is cleaner and easier to read when there might be use cases where a lot of tagging of content in a particular paragraph is required.

nobkd commented:
Yes, you need the quotation marks, though you can also use single (`'`) ones. otherwise it would be an html attribute. (though I'm not sure, if it gets rendered.)

If you have e.g. an `em`phasis around the content, it *might* work (on single words?), but i wouldn't bet on it.

nobkd committed:
feat: block and inline native html tags

nobkd commented:
Added support for native html tags using nuemark tags

Edit: Lone lists will still fail, even with the fix from #472. I might take a look in or after this one.

nobkd commented on packages/nuemark/src/render-tag.js:
```diff
@@ -142,9 +153,21 @@ export function renderIcon(name, symbol, icon_dir) {
+    // native html tags
+    if (HTML_TAGS.includes(tag.name)) {
+      // inline / block without blocks, but with '_' data
+      if (tag.is_inline || (!tag.blocks?.length && tag.data?._)) return elem(tag.name, tag.attr, renderInline(tag.data?._, opts))
```
```suggestion
      if (tag.is_inline || !tag.blocks?.length) return elem(tag.name, tag.attr, renderInline(tag.data?._, opts))
```
Could do it like this. If there's no block content, it doesn't matter if the _ data is set or not.

nobkd committed:
Simplify check for inline rendering

tipiirai commented:
Looks solid. So block tags are still rendered as divs by default?

nobkd commented:
The anonymous / unnamed block tags? Yes.

goblinfactory commented:
> Yes, you need the quotation marks, though you can also use single (`'`) ones. otherwise it would be an html attribute. (though I'm not sure, if it gets rendered.)
> 
> If you have e.g. an `em`phasis around the content, it _might_ work (on single words?), but i wouldn't bet on it.

Hi @nobkd I'm not saying it shouldnt be this way, just wanting to know why?
I dont see how it should be an html attribute, can you provide an example to show what you mean?

e.g. do you mean?

`hello [.green foo bat "*content*"]` -> `hello <span class="green" foo bat><em>content</em></span>`

or something else?
My noobie brain is wondering why it can't simply work as follows (below); this seems intuitive.

`hello [.green foo bat *content*]` -> `hello <span class="green">foo bat <em>content</em></span>`

 I'm sure I've misunderstood so just hoping you can clarify with a simple example, txs, 
A

p.s. My reaction to the "" around content, is that it suggests that CONTENT must sometimes look like programmer strings and other times must simply be content. Attributes are uncommon in content, so the default rule in my opinion should be that "content" is never quoted inside markdown, unless you're deliberately disambiguating something, e.g. a span where with attributes   (the exception, not the rule), so the `exception` becomes

`this is [.drop drop-here "*drophere*"]` -> `hello <span class="drop" drop-here><em>drophere</em></span>`
Please forgive the awful example.

This would be ugly to read inside markdown, but would leave the "normal" case much easier to read, when you simply want to [.red highlight] a particular word in a secondary or other styling that double star (**) would work.

Compared to ;

This would be ugly to read inside markdown, but would leave the "normal" case much easier to read, when you simply want to [.red "highlight"] a particular word in a secondary or other styling that double star (**) would work.

With the quotes, at a quick glance highlight looks like a special attribute, distinct from content, so that's jarring.

also, I've seen some implementations use

when you simply want to [.red]highlight[/] a particular

as an approach?

Just mentioning all of these so that we've captured the discussion somewhere and I can understand the rationale. 
Txs,
Alan

pps; I know these comments are not technically related to this PR, so apologies for that, and it should possibly? be made another issue. I'll create a new issue with a copy of this depending on the reply and answer. (If we agree it's worth discussing further)

nobkd committed:
append data attrs on html tags

nobkd commented:
@goblinfactory 

Yes, exactly. Your first example shows how it is with this pr. 

The non-quoting *could* work for anonymous / unnamed tags (though there might be some issues with parsing), but for other islands / layout components not so much, as they use mainly attributes.

I understand your reasoning, but: 
1. I just built on the existing foundation and used the parsed data
2. There can be cases where these cases would be hard to disambiguate: `[.myclass he said "hello"]` (and yes, this is definitely a crafted example) 
    - `<span class="myclass">he said <q>hello</q></span>`
    - `<span class="myclass" he said>hello</span>`

`[.class]content[/]` could be a way forward, but would have to be implemented, and could be hard to parse. (and is imo not really pretty) Would there always have to be an end tag, even though I maybe just wanted a empty span? Tags within tags could be hard to parse. (not even sure, if that currently works for inline)

I think a new issue / discussion for this would be better :)

---

Edit:
Vague idea: make your parsing the default for unnamed / anonymous tags, and keep current behavior for named ones. Something like:
- `[.blue this is all "content"]` -> `<span class="blue">this is all <q>content</q></span>`
- `[span.blue this is not all "content"]` -> `<span class="blue" this is not all>content</span>`

Not sure, if the same should be for anonymous block tags too, (if they have no children, of course)

goblinfactory commented:
txs @nobkd Lets see after this gets merged how the usage on bigger projects feels and if it looks like it's worth discussing as a priority I'll create a new issue for discussion and copy notes from here. Txs for the work. Alan

nobkd committed:
fix: unnamed block tag now defaults to div, don't skip list issue test

nobkd committed:
feat: colon yaml tag syntax

nobkd commented:
Thinking about splitting this pr in two.

One for unnamed inline spans and another one for named html tags (block and inline).

Would probably be easier to review and smaller in scope for both, even though they're tightly related. 

I'll sleep this night over it, and might open 2 new PRs tomorrow, if I still think this decision is right.

Edit: also waiting what will happen with #476

nobkd commented on packages/nuemark/src/parse-blocks.js:
```diff
@@ -182,7 +182,9 @@ function processNestedBlocks(block, capture) {
-      if (body && name && isYAML(body.trim())) {
+      // TODO: add additional check for native html tags
+      // maybe new syntax? e.g. `[yaml-tag]:\n\thi` (note colon)
+      if (body && isYAML(body.trim())) {
```
TODO: edit

nobkd committed:
use default rendering for button

nobkd committed:
docs: use new colon yaml syntax

albedosehen committed:
feat(nueserver): expand & correct MIME type support for additional web assets

albedosehen committed:
Typo in comment

albedosehen committed:
chore(test): replace import.meta.dir with Node-compatible dirname resolution

nobkd committed:
simplify test file

tipiirai committed:
Change of blog post URL for Hacker News

chrisdoram committed:
Update build-page action to align with syntax schema

Remove `type` field as is not a field defined in the metadata syntax
for github actions.

Add description to the action and `root` input parameter. The input
parameter is not required as it has a default.

chrisdoram committed:
Update all references to legacy blog entry introducing-hyper

This has been updated to `standards-first-react-alternative`
which is the valid subdirectory.

nobkd committed:
Merge pull request #574 from albedosehen/fix/web-server-mime-types

Expand MIME type map in web server to broaden asset type support with fallback (fonts, source maps, etc)

nobkd committed:
Update action descriptions

nobkd committed:
Merge pull request #578 from chrisdoram/master

Update legacy introducing-hyper link

tipiirai committed:
Form fix

nobkd committed:
deps: upgrade

nobkd committed:
fix

nobkd commented:
incremental pr on #476

nobkd committed:
try reflink fix

nobkd committed:
allow anonymous yaml tag

nobkd committed:
docs: use correct native html tags

nobkd commented:
`return renderTag(tag, opts)` causes `<div class="problem" nuekit_version="1.0.0-RC.4" globals="@global,css" libs="@lib,docs/lib,blog/lib" og="`, but without, e.g. the following code uses the link ref name as link, because it can't find the link ref inside the nuemark component:
```md
[.problem]
  [using ref][testref]

[testref]: //example.com
```

will have to fix this, while not adding all the settings, etc as attributes

nobkd committed:
fix: attributes & own data handling
